hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileBlock.java

```diff
@@ -1697,8 +1722,27 @@ protected HFileBlock readBlockDataInternal(FSDataInputStream is, long offset,
           headerBuf = HEAP.allocate(hdrSize);
           readAtOffset(is, headerBuf, hdrSize, false, offset, pread);
           headerBuf.rewind();
+
+          // The caller didn't provide an anticipated block size and headerBuf was null, this is
+          // probably the first time this HDFS block has been read. The value we just read has not
+          // had HBase checksum validation ; assume it was not protected by HDFS checksum either.
+          // Sanity check the value. If it doesn't seem right, either trigger fall-back to hdfs
+          // checksum or abort the read.
+          //
+          // TODO: Should we also check the value vs. some multiple of fileContext.getBlocksize() ?
+          onDiskSizeWithHeader = getOnDiskSizeWithHeader(headerBuf, checksumSupport);
+          if (!checkOnDiskSizeWithHeader(onDiskSizeWithHeader)) {
+            if (verifyChecksum) {
+              invalidateNextBlockHeader();
+              span.addEvent("Falling back to HDFS checksumming.", attributesBuilder.build());
+              return null;
+            } else {
+              throw new IOException("Read invalid onDiskSizeWithHeader=" + onDiskSizeWithHeader);
+            }
+          }
+        } else {
+          onDiskSizeWithHeader = getOnDiskSizeWithHeader(headerBuf, checksumSupport);
```

Comment:
i think this here is when we pull it from the cached header. is there a reason not to do the same validation as above for that?

Reply:
another thought here -- i think we can still possibly cache corrupt headers. below, we read from disk `[data block with header and checksum] + [next header]`.  we then validate the checksum on the first part, which may be correct. but we do nothing to validate the header before caching it.

im not sure we _can_ checksum just the header, so maybe we just have to live with a possibly-corrupt cache. however, we could at least validate the onDiskSizeWithHeader from that next header before caching it. so i might suggest we add one more call to getOnDiskSizeWithHeader + checkOnDiskSizeWithHeader in `cacheNextBlockHeader` maybe

Reply:
> We can, but if it is suspect we have no way to hint early that the next read must fall back to HDFS checksums. At best, the next time we read a block, we'd end up right back here without the prefetched header, do the pre-fetch again, validate again, but this time be able to do something about it.
>
> Best to always assume a pre-fetched header is suspect.

Yea good point. So we need to add a checkOnDiskSizeWithHeader here for the case where headerBuf != null

Reply:
This discussion is from an outdated portion of the PR, but I think its concern has been addressed.